src/worker/jobs/expiry_notification.rs

```diff
@@ -0,0 +1,231 @@
+use crate::models::ApiToken;
+use crate::schema::api_tokens;
+use crate::{email::Email, models::User, worker::Environment, Emails};
+use anyhow::anyhow;
+use chrono::SecondsFormat;
+use crates_io_worker::BackgroundJob;
+use diesel::dsl::IntervalDsl;
+use diesel::{
+    dsl::now, BoolExpressionMethods, ExpressionMethods, NullableExpressionMethods, PgConnection,
+    QueryDsl, QueryResult, RunQueryDsl, SelectableHelper,
+};
+use std::sync::Arc;
+
+/// The threshold in days for the expiry notification.
+const EXPIRY_THRESHOLD: chrono::TimeDelta = chrono::TimeDelta::days(3);
+
+/// The maximum number of tokens to check.
+/// To avoid sending too many emails and submitting to many transactions, we limit the number of
+/// tokens to check.
+const MAX_ROWS: i64 = 10000;
+
+/// A job responsible for monitoring the status of a token.
+/// It checks if the token is about to reach its expiry date.
+/// If the token is about to expire, the job triggers a notification.
+#[derive(Default, Serialize, Deserialize, Debug)]
+pub struct SendTokenExpiryNotifications;
+
+impl BackgroundJob for SendTokenExpiryNotifications {
+    const JOB_NAME: &'static str = "expiry_notification";
+
+    type Context = Arc<Environment>;
+
+    #[instrument(skip(env), err)]
+    async fn run(&self, env: Self::Context) -> anyhow::Result<()> {
+        let conn = env.deadpool.get().await?;
+        conn.interact(move |conn| {
+            // Check if the token is about to expire
+            // If the token is about to expire, trigger a notification.
+            check(&env.emails, conn)
+        })
+        .await
+        .map_err(|err| anyhow!(err.to_string()))?
+    }
+}
+
+// Check if the token is about to expire and send a notification if it is.
+fn check(emails: &Emails, conn: &mut PgConnection) -> anyhow::Result<()> {
+    info!("Checking if tokens are about to expire");
+    let expired_tokens = find_tokens_expiring_within_days(conn, EXPIRY_THRESHOLD)?;
+    if expired_tokens.len() == MAX_ROWS as usize {
+        warn!("The maximum number of API tokens per query has been reached. More API tokens might be processed on the next run.");
+    }
+    for token in &expired_tokens {
+        if let Err(e) = handle_expiring_token(conn, token, emails) {
+            error!(?e, "Failed to handle expiring token");
+        }
+    }
```

Comment:
There is a possibility that expired tokens are not being processed correctly. This might be due to failed tokens reaching the `MAX_ROWS` limit and being repeatedly selected, which could prevent tokens after that limit from being processed. While this should be a rare case, it's something to be aware of.

Reply:
Yes, it is possible. But it should be very rare.

Reply:
Yeah, considering the sufficiently large `MAX_ROW` value, this should be a rare case. Additionally, if these failed tokens can be identified and recovered somehow (either before processing the job or afterwards), then there shouldn't be any problems. Therefore, I don't feel strongly about requesting a change, but I just wanted to point out this potential issue.

Reply:
To fully solve this problem, we can store retry information in our database. Once we reach a certain threshold, we can ignore those tokens. But we should be able to improve it in the future if we do run into this problem.